mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/ReactorRabbitMQChannelPool.java

```diff
@@ -0,0 +1,140 @@
+        private static final Duration FOREVER = Duration.ofMillis(Long.MAX_VALUE);
+
+        private final Mono<Connection> connectionMono;
+
+        ChannelFactory(Mono<Connection> connectionMono) {
+            this.connectionMono = connectionMono;
+        }
+
+        @Override
+        public Channel create() throws Exception {
+            return connectionMono
+                .flatMap(this::openChannel)
+                .block();
+        }
+
+        private Mono<Channel> openChannel(Connection connection) {
+            return Mono.fromCallable(connection::openChannel)
+                .map(maybeChannel ->
+                    maybeChannel.orElseThrow(() -> new RuntimeException("RabbitMQ reached to maximum opened channels, cannot get more channels")))
+                .retryBackoff(MAX_RETRIES, RETRY_FIRST_BACK_OFF, FOREVER, Schedulers.elastic())
+                .doOnError(throwable -> LOGGER.error("error when creating new channel", throwable));
+        }
+
+        @Override
+        public PooledObject<Channel> wrap(Channel obj) {
+            return new DefaultPooledObject<>(obj);
+        }
+
+        @Override
+        public void destroyObject(PooledObject<Channel> pooledObject) throws Exception {
+            Channel channel = pooledObject.getObject();
+            if (channel.isOpen()) {
+                channel.close();
+            }
+        }
+    }
+
+    private static final long MAXIMUM_BORROW_TIMEOUT_IN_MS = Duration.ofSeconds(5).toMillis();
+
+    private final GenericObjectPool<Channel> pool;
+    private final ConcurrentSkipListSet<Channel> borrowedChannels;
+
+    ReactorRabbitMQChannelPool(Mono<Connection> connectionMono, int poolSize) {
+        ChannelFactory channelFactory = new ChannelFactory(connectionMono);
+
+        GenericObjectPoolConfig<Channel> config = new GenericObjectPoolConfig<>();
+        config.setMaxTotal(poolSize);
+        this.pool = new GenericObjectPool<>(channelFactory, config);
+        this.borrowedChannels = new ConcurrentSkipListSet<>(Comparator.comparingInt(System::identityHashCode));
+    }
+
+    @Override
+    public Mono<? extends Channel> getChannelMono() {
+        return Mono.fromCallable(() -> {
+            Channel channel = pool.borrowObject(MAXIMUM_BORROW_TIMEOUT_IN_MS);
+            borrowedChannels.add(channel);
+            return channel;
+        });
+    }
+
+    @Override
+    public BiConsumer<SignalType, Channel> getChannelCloseHandler() {
+        return (signalType, channel) -> {
+            borrowedChannels.remove(channel);
+            if (!channel.isOpen() || signalType != SignalType.ON_COMPLETE) {
+                invalidateObject(channel);
+                return;
+            }
+            pool.returnObject(channel);
+        };
+    }
+
+    private void invalidateObject(Channel channel) {
+        try {
+            pool.invalidateObject(channel);
+            if (channel.isOpen()) {
+                channel.close();
+            }
```

Comment:
borrowedChannels should remove that channel in here

Reply:
in this case valid `Channel`s returned to the pool would never leave `borrowedChannels`

Reply:
No, this is invalidate method, where a channel is being rejected from the pool, then it's state is rejected not borrowed, we should remove it from the borrowedChannels

Reply:
At this point it's already removed from `borrowedChannels`.

`borrowedChannels` is a set which is use to be able to close `Channel`s out of the `ObjectPool`.

When a `Channel` is returned, either it is closed or open,, it should leave `borrowedChannels` since we can act on it.

Reply:
ah, yeah, it's in the above method, sorry!